### DISCUSSION TITLE
How to integrate pip-tools to produce lock files for the Docker image's Python environment?

### DISCUSSION THREAD

**Discussion by ketozhang** [apache/airflow#37273] [Q&A]
If you go with the Docker image as a deployment, to reproduce the image's Python environment locally for development (e.g., writing DAGs), you have to reproduce a lock file after building the image and running it as a container.

```console
$ docker run --rm -it --entrypoint "bash" ${IMAGE}:${TAG}  -c \
    "echo \# `python -V` && pip freeze" \
    > requirements/requirements.lock
```

This scheme does not work well for pip-tools as the generation of the lock file is replaced by Docker and you lose a lot of benefits from pip-tools: python version, hashes, metadata comments, etc.

**Then there is dev dependencies**. Let's use a dedicated requirements file for this (instead of `pyproject.toml` since application support is quirky).

```
# requirements/dev-requirements.in
-r requirements.lock
pytest
```

Now we have to use pip-compile if we want a dev lock file:

```console
$ pip-compile requirements/dev-requirements.in -o requirements/dev-requirements.lock

# First lock file is optional to install
$ pip install -r requirements/requirements.lock requirements/dev-requirements.lock
```

**Comment by potiuk** [comment]
Airflow does not have support for tools like poetry or pip-tools - see the note https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#installation-tools the "lock" functionality is not yet standardized by the Python comunity and each frontend development tool has their own ways of implementing them and  none of them has good support for monorepo (like we have with Airflow- we have ~ 90 packages with 90 different sets of dependencies whih we maintain separately for each provider. 

Instead airflow produces constraint files (which are similar to requirements) which you can use to install the right set of dependencies for the chosen airflow version, extras and python version. You should figure out how to make it works for your workflow, often it will mean that you will have to use a combination of pip-tools and pip, or maybe simply abandon pip-tools and share with your team the way how to synchronize the dependencies. 

All the extras available for airflow (including the devel ones) are explained in https://airflow.apache.org/docs/apache-airflow/stable/extra-packages-ref.html and very recently we moved airflow to be absolutely compliant with all the recent packaging PEPs having all the dependencies described in static form in pyproject.toml https://github.com/apache/airflow/blob/main/pyproject.toml (which as of January/Airflow 2.8.0 is the single source of project dependency description) 

You can also generate your own constraints where you add your dependencies separately. The https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html explains the commands that should be used to install airflow with fixed set of dependencies using constraints.

However, since you are pip-tools user, if you figure out a nice way to integreate our constraints it and have a good workflow, we are absolutely open to accept a PR describing it in our "Installation from PyPI".

**Reply by ketozhang** [reply]
Thank you! I have indeed looked very closely at the constraints and user-defined constraints section before this discussion. (PS: with the new pip resolver, do try out the user-defined constraints; I don't think it works anymore).

I do have a way for pip-tools to integrate with non-image installs :). I'd be happy to add a PR to the docs for this and we can have a discussion there. However, as I've mentioned here. Docker Image installs fairly awkward, so I'm afraid the docs won't be as useful for users doing image deployments who hope to reproduce the image environment on their local dev machine.

The crux of this issue is there's a difference where the origin of the dependency spec come from. In a non-image install the dependency spec comes from `pip install apache-airflow[extras]` + `constraints.txt`. For image installs, the dependency spec comes from the image itself. In the first case the user has full control over the deps and constraints, in the latter case the user must rectify with the existing installed packages in the image.

**Reply by potiuk** [reply]
No. you get it backwards. pip install + constraints are used to create the image.  The set of extras chosen + constraints are the source of truth, not the other way.  This shoud be the same for any other way of installing airflow deps. 

So both non-image and image install come from the same source.

Unless I completely misunderstood what your "dependency spec comes from the image itself".

Image is the result, not the source of installing dependencies.

**Reply by potiuk** [reply]
>  (PS: with the new pip resolver, do try out the user-defined constraints; I don't think it works anymore). 

I believe they work. there were no changes in behaviour, I am following it very closely. But if you get some examples of what you are talking about,  maybe I missed something. Would you like to elaborate what do you mean b y "I don't think it works anymore"? Courious to hear it.

**Reply by ketozhang** [reply]
Ignore me about the constraints not working. This is an issue with pip-compile using Python 3.12 (which Airflow 2.8 does not support). Indeed `pip` works if it's part of Python<3.12 .

**Reply by ketozhang** [reply]
> No. you get it backwards.

I'm talking about the perspective of the user not Airflow package & image builder. From the user's (installer's) perspective, using the image, the `pip install` and constraints are hidden from the user. Think of it this way. If my goal is to create a lock file, what are my inputs? For the non-image, it's pip install and constraints; for image, it is the `docker run ... pip freeze`.

**Reply by potiuk** [reply]
No. your source is always airflow package. Image is a product not source. You are choosing a wrong source of truth

**Reply by ketozhang** [reply]
Let's say I chose the airflow package and constraints as the source of truth for my lock file. Does that guarantee that image docker environment and the lock file to be the same? 

Remember, the goal is to produce a lock file that can be used in development on a local machine which accurately represents the deployed environment (the image)*

*Let's ignore developing inside a container. Definitely an option with modern tools, but adoption is slow.

**Reply by potiuk** [reply]
If you use the lock file to build the image, yes, they will be the same

**Reply by ketozhang** [reply]
Oh I see where we're conflicting.

I am also limiting the case to building the image with Airflow as a base image. Rarely do users opt for starting from scratch because it's very difficult to integrate it with existing deployment methods (e.g., Helm chart) since there's some standards made in the images (e.g., /opt/airflow is home)

So it looks like I have two choices:
- Start with a Python base image and use a lock file to install Airflow. Then try to be compatible with existing image deployment.
- Start with an Airflow base image and forgo the lock file.

**Reply by potiuk** [reply]
Well `pip-tools` is very opinionated in the way it does things. It has it's own way of approaching the depenedencies. Neither requirements.in nor (any more) requirements.txt) are standard nor even recommended way of instaling python packages. While there is no PEP describing the lock-like file behaviour yet, there are discussions on how to standardize it https://peps.python.org/pep-0735/ - draft PEP that describes dependency groups. 

You can likely write your script that will generate requirements.in file for you - that would be the best way until PEP-735 gets approved and implemented by various tools (it's mentioned there that it will be great source to generate lock files).
